arch/x86/core/early_serial.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <generated_dts_board.h>
+
+/* Super-primitive 8250/16550 serial output-only driver, 115200 8n1 */
+
+#define _PORT DT_UART_NS16550_PORT_0_BASE_ADDR
+
+static inline void ioport_out8(unsigned short port, unsigned char b)
+{
+	__asm__ volatile("outb %0, %1;\n\t" : : "a"(b), "d"(port));
+}
+
+static inline unsigned char ioport_in8(unsigned short port)
+{
+	unsigned char ret;
+
+	__asm__ volatile("inb %1, %0;\n\t" : "=a"(ret) : "d"(port));
+	return ret;
+}
+
+static void serout(int c)
+{
+	while (!(ioport_in8(_PORT + 5) & 0x20)) {
```

Comment:
I hate the naked constants here.  In the future maybe we could move this into the ns16550 driver file (just as some #ifdef'd functions that are like "here's the standalone stuff for early console") so we can use the definitions there, plus keep everything 16550-related in one place.

Reply:
I'll punt on 16550 driver unification, but I agree about naked constants, will fix